jax/experimental/pjit.py

```diff
@@ -0,0 +1,211 @@
+def pjit(fun: Callable,
+         in_mesh_axes,
+         out_mesh_axes,
+         static_argnums: Union[int, Iterable[int]] = (),
+):
+  warn("pjit is an experimental feature and probably has bugs!")
+  _check_callable(fun)
+
+  # To be a tree prefix of the positional args tuple, in_axes can never be a
+  # list: if in_axes is not a leaf, it must be a tuple of trees. However,
+  # in cases like these users expect tuples and lists to be treated
+  # essentially interchangeably, so we canonicalize lists to tuples here
+  # rather than raising an error. https://github.com/google/jax/issues/2367
+  if isinstance(in_mesh_axes, list):
+    in_mesh_axes = tuple(in_mesh_axes)
+  if isinstance(out_mesh_axes, list):
+    out_mesh_axes = tuple(out_mesh_axes)
+
+  static_argnums = _ensure_index_tuple(static_argnums)
+
+  @wraps(fun)
+  def wrapped(*args, **kwargs):
+    if kwargs:
+      raise NotImplementedError("pjit over kwargs not yet supported")
+
+    # Putting this outside of fun_mapped would make resources lexically scoped
+    resource_env = maps.thread_resources.env
+
+    f = lu.wrap_init(fun)
+    if static_argnums:
+      if max(static_argnums) >= len(args):
+        raise ValueError(
+            f"jitted function has static_argnums={static_argnums}"
+            f" but was called with only {len(args)} positional "
+            f"argument{'s' if len(args) > 1 else ''}. "
+            "All static broadcasted arguments must be passed positionally.")
+      dyn_argnums = [i for i in range(len(args)) if i not in static_argnums]
+      f, args = argnums_partial(f, dyn_argnums, args)
+
+    args_flat, in_tree = tree_flatten(args)
+    for arg in args_flat: _check_arg(arg)
+    flat_fun, out_tree = flatten_fun_nokwargs(f, in_tree)
+
+    in_mesh_axes_flat = tuple(flatten_axes("pjit in_mesh_axes",
+                                           in_tree, in_mesh_axes))
+    out_mesh_axes_thunk = HashableFunction(
+        lambda: tuple(flatten_axes("pjit out_mesh_axes", out_tree(), out_mesh_axes)),
+        closure=out_mesh_axes)
+
+    out = pjit_call_p.bind(
+        flat_fun,
+        *args_flat,
+        in_mesh_axes=in_mesh_axes_flat,
+        out_mesh_axes_thunk=out_mesh_axes_thunk,
+        resource_env=resource_env,
+        name=flat_fun.__name__)
+    return tree_unflatten(out_tree(), out)
+
+  return wrapped
+
+def _pjit_call_impl(fun: lu.WrappedFun, *args, in_mesh_axes,
+                    out_mesh_axes_thunk, resource_env, name):
+  in_avals = [core.raise_to_shaped(core.get_aval(arg)) for arg in args]
+  return _pjit_callable(
+      fun, in_mesh_axes, out_mesh_axes_thunk, resource_env, name,
+      *in_avals)(*args)
+
+def _pjit_translation_rule(c, axis_env, in_nodes, name_stack, backend, name,
+                           call_jaxpr, in_mesh_axes, out_mesh_axes_thunk,
+                           resource_env):
+  # TODO(skye): this should arguably be part of axis_env
+  mesh = resource_env.physical_mesh
```

Comment:
I don't think it should. Adding those axes to `axis_env` would mean that the collective lowering rules could pick them up, but `pjit` doesn't introduce any axis names into scope!

Reply:
We currently do have an invariant that you can't switch the `resource_env` on any primitive wrapped in `xmap`/`pjit`, so I think it's fine. I'd just rather keep those mesh axes out of `axis_env`, because if we ever end up accidentally calling into e.g. a `psum` rule then it might succeed for no good reason instead of throwing an unbound axis error.